text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,592 @@
+| `deliver`      | One or more messages are passed to a consumer. |
+| `settle`       | One or more messages are settled. |
+
+For further details about each of those operations refer to the [section about trace structure](#trace-structure).
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+
+The kind of `settle` spans should be set according to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
+e. g. it should be set to `CLIENT` if the `settle` spans models a synchronous call
+to the intermediary.
+
+Setting span kinds according to this table ensures that span links between
+consumers and producers always go from a `PRODUCER` span on the producer side to
+a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
+linked traces without the need for additional semantic hints.
+
+### Span relationships
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
+SHOULD account only for a single message.
+
+If a "Create" span exists for a message, its context SHOULD be injected into
+the message. If no "Create" span exists, the context of the related "Publish"
+span SHOULD be injected into the message.
+
+#### Consumer spans
+
+##### Push-based scenarios
+
+"Deliver" spans SHOULD be created for operations of passing messages to the
+application when those operations are not initiated by the application
+code.
+
+##### Pull-based scenarios
+
+"Receive" spans SHOULD be created for operations of passing messages to the
+application when those operations are initiated by the application code.
+
+##### General considerations
+
+"Deliver" or "Receive" spans MUST NOT be created for messages which are not
+forwarded to the caller, but are pre-fetched or cached by messaging
+libraries or SDKs.
+
+A single "Deliver" or "Receive" span can account for a single message, for
+multiple messages (in case messages are passed for processing as batches), or
+for no message at all (if it is signalled that no messages were received).  For
+each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
+the "Create" or "Publish" span for the message. In addition, if it is possible
+the "Create" or "Publish" span MAY be set as a parent of the "Deliver" or
```

Comment:
Can we clarify this for batch-receive, where multiple messages/producer spans "fan in" to a single receive span? There was some dicussion around that for AWS Lambda SQS at https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1051#issuecomment-1492251707

IMHO: For a batch receive, you may still want to set any of the producer spans as parent e.g.:
* One may argue that the last message in the batch is the one that caused(!) the receive operation to finish since it filled up the batch.
* Semantically, links may imply a much weaker causal relationship than parent-child relationships. But for message processing, if not for batches, parent-child would often be quite appropriate.
* For technical reasons, it may be helpful to also have a parent: Links are usually not as convenient to use and process in backends. For example, links may be shown as actual clickable links only, while parent-childs may be shown in a single tree. With parent-child also for batches, you will be able to find an exemplar (parent-child) trace for any particular path through your system given enough samples (i.e. traces).

Reply:
What you describe might theoretically work for "deliver" cases (push-based scenarios), as in those cases the batch of messages is already ready before the "deliver" span is started. Disregarding internals like message pre-fetching and caching, it's not possible for "receive" cases (pull-based scenarios). In those cases, we must assume that no message is present when the "receive" operation (and span) starts, and messages are fetched during this "receive" operation. Thus, no parent producer context is available and can be used when the "receive" span starts.

> * One may argue that the last message in the batch is the one that caused(!) the receive operation to finish since it filled up the batch.

When talking about causality in the context of parent/child relationships, this usually implies causation for _starting_ an operation, but not for _finishing_ it.

> * Semantically, links may imply a much weaker (non-)causal relationship than parent-child relationships. But for message processing, if not for batches, parent-child would often be quite appropriate.

For single-message cases, this OTEP already recommends creating a parent/child relationship.

> * For technical reasons, it may be helpful to also have a parent: Links are usually not as convenient to use and process in backends. For example, links may be shown as actual clickable links only, while parent-childs may be shown in a single tree. With parent-child also for batches, you will be able to find an exemplar (parent-child) trace for any particular path through your system given enough samples (i.e. traces).

This OTEP doesn't forbid instrumentors to select any message creation context as parent in batch consumer scenarios. If this is considered beneficial for certain scenarios (e. g. SQS), it can be done while still being conforming with this proposal.

However, I think it shouldn't be generally recommended or mandated for all cases. There are many sources of ambiguity and possible confusion that this could cause:
* How would you interpret the trace duration in such a case? You'd end up with a trace that at some points starts with producing message A, but then also includes the time needed for batch-consuming multiple other, possibly unrelated messages. The trace duration doesn't tell you anything anymore about message A: maybe message A is processed very quickly, but another message in the batch takes very long to process?
* Interpreting errors in traces becomes trickier too. You might have a trace started with producing message A, but an error could happen during batch-consuming of another message.
* Also, often consuming message cause others to be produced. You might start out with producing message A, a batch-consuming operation involving message A and message B is parented, during consuming message B, message C is produced, and a batch-consuming operation involving message C is parented to the trace that started with producing message A, although this operation has nothing at all to do with it.

The case of examplar traces is an interesting one, however, when aiming at creating meaningful traces that track a single message, parenting based on a single message in batch-scenarios can be misleading, and even harmful.

Reply:
You bring up good points. To respond to just one of them: in fact the example I had in mind is probably called a "deliver" span in this spec, despite me using the wrong term "receive".

Reply:
Re: Trace duration: I wonder if the combination of "span with kind produce" and child has IsRemote (cf. OTEP #182) could/should be used to partition a trace into subtrees for which it is more sensible to calculate duration totals. But I need to read up on how process spans are linked/parented relative to deliver spans...

Reply:
> "span with kind produce" and child has IsRemote

That could work for some scenarios, but not for batch-consuming ones.

The consistent logic (across all scenarios) with the conventions proposed in this OTEP would be _"span with kind producer" and "linked span(s) of type consumer"_.

Reply:
@Oberon00 Considering that there is a separate issue that will cover adding semantic conventions for "Process" spans, do you think that this conversation can be closed here and this OTEP unblocked?